sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+from sbi.utils.torchutils import ensure_theta_batched
+from sbi.utils.user_input_checks import process_x
+from sbi.inference.posteriors.base_posterior import NeuralPosterior
+from sbi.inference.potentials.base_potential import BasePotential
+
+from torch import Tensor
+from typing import List, Optional, Union, Any
+from sbi.types import Shape, TorchTransform
+
+
+class NeuralPosteriorEnsemble(NeuralPosterior):
+    r"""Wrapper class for bundling together different posterior instances into an
+    ensemble.
+
+    This class creates a posterior ensemble from a set of N different, already trained
+    posterior estimators $p_{i}(\theta|x_o)$, where $i \in \{i,...,N\}$.
+
+    It can wrap all posterior classes available in sbi and even a mixture of different
+    posteriors, i.e. obtained via SNLE and SNPE at the same time, since it only
+    provides a pass-through to the class-methods of each posterior in the ensemble. The
+    only constraint is, that the individual posteriors have the same prior.
+
+    So far `log_prob()`, `sample()` and `map()` functionality are supported.
+
+    Example:
+
+    ```
+    import torch
+    from joblib import Parallel, delayed
+    from sbi.examples.minimal import simple
+
+    ensemble_size = 10
+    posteriors = Parallel(n_jobs=-1)(delayed(simple)() for i in range(ensemble_size))
+
+    ensemble = NeuralPosteriorEnsemble(posteriors)
+    ensemble.set_default_x(torch.zeros((3,)))
+    ensemble.sample((1,))
+    ```
+
+    Attributes:
+        posteriors: List of the posterior estimators making up the ensemble.
+        num_components: Number of posterior estimators.
+        weights: Weight of each posterior distribution. If none are provided each
+            posterior is weighted with 1/N.
+        potential_fn: Potential function of the ensemble.
+        prior: Prior distribution that is the same for all posteriors. If it is not the
+            same, then `prior = None`.
+        device: Device which the component distributions sit on.
+        default_x: Used in `.sample(), .log_prob()` as default conditioning context.
+    """
+
+    def __init__(
+        self,
+        posteriors: List,
+        weights: Optional[Union[List[float], Tensor]] = None,
+        theta_transform: Optional[TorchTransform] = None,
+    ):
+        """
+        Args:
+            posteriors: List containing the trained posterior instances that will make
+                up the ensemble.
+            weights: Assign weights to posteriors manually, otherwise they will be
+                weighted with 1/N.
+            theta_transform: If passed, this transformation will be applied during the
+            optimization.
+        """
+        self.posteriors = posteriors
+        self.num_components = len(posteriors)
+        self.weights = weights
+        self.potential_fn = None  # will be set when context is provided.
+
+        self.device = self.ensure_same_device(posteriors)
+
+        self.prior = self.common_prior()
```

Comment:
This is tricky. Does this method work with all priors supported by `sbi`? E.g. `MultipleIndependent` or scipy distributions wrapped as torch dists? If not, I would suggest that we remove this test (ideally in a single commit so that we could revert). I'm perfectly fine without this test, it is well documented in the docstring above anyways. It's also kind of common sense that they should all have the same prior.

Reply:
I do not know that for sure, since I have not tried it for all manner of distributions. But I tried to implement something akin to __eq__() and I am not just comparing class membership / type, but their attributes on a one by one basis. I think this should work or at least be extensible to the other distributions that are around. Might have to do some type comparison adjustments ala:
if type ndarray:
np.compare(dist_x, dist_y)
if type sh_else:
sth_else.compre(dist_x, dist_y)

Reply:
I'm voting for not comparing them at all. If you compare them, you definitely need to make sure (and add tests) that it works with all priors supported by sbi. Again, I vote for dropping it.

Reply:
I think I have an idea how I could ensure_same_prior for all priors.
Something like
```python

theta_x = dist_x.sample((n,))
for dist_y in component_priors:
   assert dist_y.log_prob(theta_x) == dist_x.log_prob(theta_x), "priors are not the same'
```

Since every prior is required to have a `.sample()` and `.log_prob()`, this should work I believe.

Reply:
yup that should work